File: 239-滑动窗口最大值/239.cpp
class Solution {
public:
    // 本题的解法可以使用双向队列
    // 首先申明一个deque 用来存储每一次滑动的大小为k的窗口中最大值在所给的数组中的下标
    // 再最后从deque中拿到所有的最大值放入定义好的vector中返回
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq;
        vector<int> vec;
        int n = nums.size();
        if(k>n || n==0){
            return vec;
        }
        for(int i = 0;i<n;i++){
            if(!dq.empty()){
                if(i>=dq.front()+k){
                    dq.pop_front();
                }
                while(!dq.empty() && nums[i]>nums[dq.back()]){
                    dq.pop_back();
                }
            }
            dq.push_back(i);
            if(i+1>=k){
                vec.push_back(nums[dq.front()]);
            }
        }
        return vec;
    }
};

File: 239-滑动窗口最大值/239.java
class Solution {
  public int[] maxSlidingWindow(int[] nums, int k) {
      Deque<Integer> queue = new LinkedList<>();
      int[] res = new int[nums.length - k + 1];
      int index = 0;
      for(int i = 0;i<nums.length;i++){
          while(!queue.isEmpty() && nums[queue.peekLast()] <= nums[i]){
              queue.pollLast();
          }
          queue.addLast(i);
          if(queue.peekFirst() == i - k){
              queue.pollFirst();
          }
          if(i >= k - 1){
              res[index++] = nums[queue.peekFirst()];
          }
      }
      return res;
  }
}